tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow.training.optimizers;
+
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Operation;
+import org.tensorflow.Output;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.core.Assign;
+import org.tensorflow.op.core.NoOp;
+import org.tensorflow.op.core.Variable;
+import org.tensorflow.types.family.TType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ *
+ */
+public abstract class Optimizer {
+  public static final String VARIABLE_V2 = "VariableV2";
+
+  /**
+   * Top level map key is the variable name, lower level map key is the slot name.
+   */
+  private final Map<String, Map<String, Variable<?>>> slots;
+
+  /**
+   * Global state variables
+   */
+  //TODO make this be used.
+  protected final List<Variable<?>> globals;
+
+  /**
+   * The Graph this optimizer is operating on.
+   */
+  protected final Graph graph;
+
+  /**
+   * The ops builder for the graph.
+   */
+  protected final Ops tf;
+
+  protected Optimizer(Graph graph) {
+    this.graph = graph;
+    this.tf = Ops.create(graph).withName(getOptimizerName());
```

Comment:
would it be easy to allow the user override the name of the scope for the optimizer ops (instead of `getOptimizerName()`)? Also, maybe a user would like to pass its own instance of `Ops` (which might already be a subscope of another block or contains control dependencies, etc.)?

Reply:
It's locked off to `Graph` at the moment as I don't think any of this works in eager mode, so the type system enforces that. I can allow a hook for the name, and we could relax the `Graph` check to an `instanceof` check throwing `IllegalArgumentException` to supply an ops. Should I keep the Graph entry point as well?

Reply:
Yes I think it is ok that to have an entry point that accept a `Graph` but maybe add another that accept an `Ops` as well, so the user has full control on the name and control dependencies of the optmizers? Or maybe just a `String` for the name

For instance, does it make sense that a user would want to create two different optimizers of the same type (but with different parameters) for handling different variables? If so, then he'll need to give them different names or he will endup with a name conflict.

I think you understand my questioning here, I'll let you decide what would be the best approach to handle those corner cases.

Reply:
I added an additional constructor which accepts a Graph and a String for the base name of the operations.